JENKINS-43637 Removed blacklisting from PostBuildScript plugin

dheid opened the pull request:
and customized update center warning as recommended by @daniel-beck here:

https://github.com/jenkins-infra/repository-permissions-updater/pull/492#issuecomment-341496738

daniel-beck commented on src/main/resources/warnings.json:
```diff
@@ -364,9 +364,11 @@
+		{
+			"firstVersion": "0.1",
+			"lastVersion": "0.17",
+			"pattern": "\\b0\\.(\\d|1[1-7])\\b"
```
This isn't matching substrings, so `\\b` are unnecessary.

For consistency, please use `[.]` for literal period chars.

We also generally try to allow for `1.2.3-whatever-3` version numbers, check other version regexes for how to allow for that.

daniel-beck commented on src/main/resources/warnings.json:
```diff
@@ -364,9 +364,11 @@
-        "pattern": ".*"
-      }
+		{
+			"firstVersion": "0.1",
```
If this is the first ever release, please leave this out.

dheid commented:
Thanks, @daniel-beck . I realized the changes and did another commit.

The changes apply to version 0.18 with compatibility to previous versions. You'll find this version here: https://github.com/jenkinsci/postbuildscript-plugin/tree/postbuildscript-0.18

I'm working on a new version 1.0.0 on the master branch. Since the users requested to have multiple scripts depending on the build result, I introduced a target result field to each script item. This requires a new configuration scheme and so there will be no compatibility to older versions. But that is not within the scope of this pull request.

daniel-beck commented:
> This requires a new configuration scheme and so there will be no compatibility to older versions. But that is not within the scope of this pull request.

Ideally you'll be able to migrate existing configurations to their equivalent in the new scheme. I'd be happy to take a look and recommend a strategy.

For example, I recently changed an existing boolean option (checkbox) to a selection of one of several options in matrix-auth without breaking data compatibility: https://github.com/jenkinsci/matrix-auth-plugin/pull/22

While as maintainer, you choose the way you want to go, but the Jenkins project values compatibility: https://jenkins.io/project/governance/#compatibility-matters

dheid commented:
Good input! I updated the plugin to migrate existing configurations from previous versions.

dheid commented:
> Still need to take a look at the plugin. ETA ~weekend.

Any news on this, @daniel-beck ? The users want their new features. ;-)

dheid commented:
Is there any way how I can support the review?

The relevant commit is https://github.com/jenkinsci/postbuildscript-plugin/commit/a6e82abb5c5be6b5303f548bffc7ebbe9ae27d76 The plugin creates a SecureGroovyScript object to evaluate the users script (previously a simple String). This isn't the most comfortable for the users, since environment variables will be replaced first and therefore the resulting script must be acknowledged each time. But by this way, you cannot inject dangerous Groovy scripts using variables.

Other commits are

* Maven POM dependency removals and upgrades: https://github.com/jenkinsci/postbuildscript-plugin/commit/3b4172f8d6d27b3440d4a06909abca517163bb61 We don't need to name the specific project types that are supported one by one and add a dependency to them. We simply think, that the plugin will work. The most common project types (Freestyle, Maven, ...) will work. If there are problems with project types, we can exclude them or fix the problems.
* https://github.com/jenkinsci/postbuildscript-plugin/commit/71e3782ea4222a62203ab3813edce50880189694 allows parameters . The plugin now splits the command into several parts and the first part has to be an existing file.
* A lot of the commits are README enhancements, formatting conventions, CI integration, find bugs issues and reformattings: https://github.com/jenkinsci/postbuildscript-plugin/commit/d467f0860b7bbb40fbb5ef173508bb52f75693d9 https://github.com/jenkinsci/postbuildscript-plugin/commit/871b35a947bb7e77f94c1b9ee601aa9f4217c8f8 https://github.com/jenkinsci/postbuildscript-plugin/commit/2e185672afdb12b7d986630629ec4ee117a7c466 https://github.com/jenkinsci/postbuildscript-plugin/commit/ebdf92b38416dd287d054a7f2a5212964f2c594f https://github.com/jenkinsci/postbuildscript-plugin/commit/d5476ee02c3b25b6ab5235ad65369928cb87cad0 https://github.com/jenkinsci/postbuildscript-plugin/commit/7a0ed0e9894226158fd61452723504b20528fbf4 https://github.com/jenkinsci/postbuildscript-plugin/commit/1ef95e4e00d67eec6eb7398e3705550ffee51141 https://github.com/jenkinsci/postbuildscript-plugin/commit/99625ec6f80251a8bd446711986fe20120bbd585 https://github.com/jenkinsci/postbuildscript-plugin/commit/dc924097b37678df5ca2db08127dc2527abdeeab https://github.com/jenkinsci/postbuildscript-plugin/commit/3d3c79e083a0764e1ed6c48811d0da38c47c976a

And that's it. :-)

rhodrid commented:
Would be great to get this de-listed as we're still building our images with the old version.

daniel-beck commented:
Review/testing notes:

* I don't understand why the callables are `SlaveToMaster…`. They seem like they should be `MasterToSlave…`. Added in https://github.com/jenkinsci/postbuildscript-plugin/commit/3b4172f8d6d27b3440d4a06909abca517163bb61 without further notes.

Some further random notes:

* No sandboxing support is weird. It basically precludes use of simple scripts in a sandbox, but not a security concern.
* Using the somewhat recent `MatrixAggregatable` extension point, the dependency on `matrix-project` could probably be made optional.
* "Add a Groovy script content" is ungrammatical.

daniel-beck commented:
Relevant diff: https://github.com/jenkinsci/postbuildscript-plugin/compare/postbuildscript-0.17...postbuildscript-0.18

dheid commented:
Thanks for the review!

I was a little bit confused by the slave-to-master and master-to-slave concepts and did a mistake. Fixed it: https://github.com/jenkinsci/postbuildscript-plugin/commit/0c53aa5e9ebbacd1e140b2ed87c9c42c26669d8e

For the sandboxing, I opened a new Jira ticket: https://issues.jenkins-ci.org/browse/JENKINS-48014

Sorry, but I didn't get how to use MatrixAggregatable without the matrix-project dependency. Do you have a corresponding web link for me to read?

I also fixed the wrong grammar: https://github.com/jenkinsci/postbuildscript-plugin/commit/0dbcc143b7c7f4b0fb207f3b6f359e6246fc8c14

The new version 0.18.1 with the fixes is just released.

daniel-beck commented:
@dheid Sorry, should have indicated the notes above were just an intermediate review result. 

> Sorry, but I didn't get how to use MatrixAggregatable without the matrix-project dependency. Do you have a corresponding web link for me to read?

You still have the dependency, but `@Extension(optional = true) public class YourMatrixAggregatable extends MatrixAggregatable` and `<optional>true</optional>` on the dependency on the POM, with some other careful coding, isolates the code touching on Matrix Project from the rest of the plugin in a way that allows use of the plugin without `matrix-project` installed.

dheid commented:
Okay, I think I got it and isolated Matrix dependency into an own post build step:

https://github.com/jenkinsci/postbuildscript-plugin/commit/6c7f614774ab467187644b4a181424f272c72630

But I did that on the most recent, refactored version of the script. Not yet in 0.18.x

daniel-beck commented:
@dheid Right, as per above, that is unrelated to this request, just something I wondered about since you were getting rid of other dependencies. Please note that testing optional dependencies isn't straightforward (I believe a `@LocalData` test could do it with `plugins/matrix-project.jpi.disabled` provided, but haven't tried it), so this could backfire.

daniel-beck commented:
I see `GroovyScriptExecutionCallable` was deleted in https://github.com/jenkinsci/postbuildscript-plugin/commit/197c0de875172dc236245bd98b800ca67e0c85a0, so its `SlaveToMasterCallable` isn't a problem anymore either.

daniel-beck commented:
I am reasonably confident this isn't terribly broken so will restore distribution.

The binding of `workspace` to a `File` seems weird given that there's apparently plans to introduce sandboxing (and basically all of `File` is a terrible choice for whitelisting), but that's not currently a security issue. Just something to keep in mind.

dheid commented:
Okay, thanks for the feedback and the merge. I'll think about a solution for the sandboxing and whitelisting issue.
